src/draw/lv_image_decoder.c

```diff
@@ -72,106 +93,56 @@ void _lv_image_decoder_deinit(void)
 {
     lv_memzero(dsc, sizeof(lv_image_decoder_dsc_t));
 
     if(src == NULL) return LV_RESULT_INVALID;
-    lv_image_src_t src_type = lv_image_src_get_type(src);
-    if(src_type == LV_IMAGE_SRC_VARIABLE) {
-        const lv_image_dsc_t * img_dsc = src;
-        if(img_dsc->data == NULL) return LV_RESULT_INVALID;
-    }
-
-    dsc->src_type = src_type;
+    dsc->src = src;
+    dsc->src_type = lv_image_src_get_type(src);
 
-    if(dsc->src_type == LV_IMAGE_SRC_FILE) {
-        dsc->src = lv_strdup(src);
-    }
-    else {
-        dsc->src = src;
-    }
+#if LV_CACHE_DEF_SIZE > 0
+    dsc->cache = img_cache_p;
+    /*
+     * Check the cache first
+     * If the image is found in the cache, just return it.*/
+    if(try_cache(dsc) == LV_RESULT_OK) return LV_RESULT_OK;
+#endif
 
-    lv_result_t res = LV_RESULT_INVALID;
+    /*Find the decoder that can open the image source, and get the header info in the same time.*/
+    dsc->decoder = image_decoder_get_info(src, &dsc->header);
+    if(dsc->decoder == NULL) return LV_RESULT_INVALID;
 
-    lv_image_decoder_t * decoder;
+    /*Duplicate the source if it's a file*/
+    if(dsc->src_type == LV_IMAGE_SRC_FILE) dsc->src = lv_strdup(dsc->src);
 
-    static const lv_image_decoder_args_t def_args = {
+    /*Make a copy of args*/
+    dsc->args = args ? *args : (lv_image_decoder_args_t) {
         .stride_align = LV_DRAW_BUF_STRIDE_ALIGN != 1,
         .premultiply = false,
         .no_cache = false,
         .use_indexed = false,
     };
 
-    /*Make a copy of args */
-    dsc->args = args ? *args : def_args;
-
-    _LV_LL_READ(img_decoder_ll_p, decoder) {
-        /*Info and Open callbacks are required*/
-        if(decoder->info_cb == NULL || decoder->open_cb == NULL) continue;
-
-        res = decoder->info_cb(decoder, src, &dsc->header);
-        if(res != LV_RESULT_OK) continue;
-
-        if(dsc->header.stride == 0) dsc->header.stride = img_width_to_stride(&dsc->header);
-
-        dsc->decoder = decoder;
-
-#if LV_CACHE_DEF_SIZE > 0
-        dsc->cache = img_cache_p;
-
-        /*Check the cache first*/
-        if(try_cache(dsc) == LV_RESULT_OK) return LV_RESULT_OK;
-#endif
-
-        res = decoder->open_cb(decoder, dsc, args);
-
-        /*Opened successfully. It is a good decoder for this image source*/
-        if(res == LV_RESULT_OK) return res;
-
-        /*Prepare for the next loop*/
-        lv_memzero(&dsc->header, sizeof(lv_image_header_t));
-
-        dsc->error_msg = NULL;
-        dsc->decoded  = NULL;
-        dsc->cache_entry = NULL;
-        dsc->user_data = NULL;
-        dsc->time_to_open = 0;
-    }
-
-    if(dsc->src_type == LV_IMAGE_SRC_FILE)
-        lv_free((void *)dsc->src);
+    /*
+     * We assume that if a decoder can get the info, it can open the image.
+     * If decoder open failed, free the source and return error.
+     * If decoder open succeed, add the image to cache if enabled.
+     * */
+    lv_result_t res = dsc->decoder->open_cb(dsc->decoder, dsc);
```

Comment:
where to free `dsc->src` if src is LV_IMAGE_SRC_FILE ?
should each `open_cb`  of image decoder do it?  I don't see any place do that.

Reply:
Regardless of success or failure, decoder `open` and `close` appear in pairs, with src allocated in `open` and released in `close`.

Reply:
Will you move the unified `cache_free_cb` to lv_image_decoder.c?

Reply:
Haven't yet.

Create a new pr to deal this thing separately. 😁

I haven't forgotten it. It's on my to-do list.